wolfcrypt/src/port/nxp/se050_port.c

```diff
@@ -0,0 +1,1418 @@
+static sss_key_store_t *hostSession;
+int keyId_allocater = 100;
+
+
+int wolfcrypt_se050_SetConfig(sss_session_t *pSession, sss_key_store_t *pHostSession) 
+{
+	printf("Setting SE050 session configuration\n");
+	
+	XMEMSET(&cfg_se050_i2c_pi, 0, sizeof(cfg_se050_i2c_pi));
+	cfg_se050_i2c_pi = pSession;
+	
+    XMEMSET(&hostSession, 0, sizeof(hostSession));
+	hostSession = pHostSession;
+    
+	return 0;
+}
+
+int se050_allocate_key()
+{
+    //keyId_allocater = rand();
+    //return keyId_allocater;
+    keyId_allocater++;
+    return keyId_allocater++;
+}
+
+
+int se050_get_random_number(uint32_t count, uint8_t* rand_out)
+{
+    sss_status_t status;
+	sss_rng_context_t rng;
+
+
+    if (wolfSSL_CryptHwMutexLock() == 0) { 
+        status = sss_rng_context_init(&rng, cfg_se050_i2c_pi);
+        if (status != kStatus_SSS_Success) {
+            return -1;
+        }
+
+        status = sss_rng_get_random(&rng, rand_out, count);
+        if (status != kStatus_SSS_Success) {
+	        return -1;
+        }
+
+        status = sss_rng_context_free(&rng);
+        if (status != kStatus_SSS_Success) {
+            return -1;
+        }
+    }
+    wolfSSL_CryptHwMutexUnLock();
+    printf("random ran correctly\n");
+    return 0;
+}
+
+
+int se050_hash_init(SE050_HASH_Context* se050Ctx, void* heap)
+{
+    se050Ctx->heap = heap;
+    se050Ctx->len  = 0;
+    se050Ctx->used = 0;
+    se050Ctx->msg  = NULL;
+    return 0;
+}
+
+int se050_hash_update(SE050_HASH_Context* se050Ctx, const byte* data, word32 len)
+{
+    if (se050Ctx == NULL || (len > 0 && data == NULL)) {
+        return -1;
+    }
+    
+    if (se050Ctx->len < se050Ctx->used + len) {
+        if (se050Ctx->msg == NULL) {
+            se050Ctx->msg = (byte*)XMALLOC(se050Ctx->used + len, se050Ctx->heap,
+                    DYNAMIC_TYPE_TMP_BUFFER);
+        } 
+        else {
+            byte* pt = (byte*)XREALLOC(se050Ctx->msg, se050Ctx->used + len, se050Ctx->heap,
+                    DYNAMIC_TYPE_TMP_BUFFER);
+            if (pt == NULL) {
+                return -1;
+            }
+            se050Ctx->msg = pt;
+        }
+        se050Ctx->len = se050Ctx->used + len;
+    }
+    XMEMCPY(se050Ctx->msg + se050Ctx->used, data , len);
+    se050Ctx->used += len;
+    return 0;
+}
+
+int se050_hash_final(SE050_HASH_Context* se050Ctx, byte* hash, size_t digestLen, sss_algorithm_t algo)
+{
+    sss_status_t status;
+    sss_digest_t digest_ctx;
+    XMEMSET(&digest_ctx, 0, sizeof(digest_ctx));
+    
+    const byte*     data = se050Ctx->msg;
+    int             size = (se050Ctx->len) / SSS_BLOCK_SIZE;
+    int             leftover = (se050Ctx->len) % SSS_BLOCK_SIZE;
+    const byte*     blocks;
+                    blocks = data;    
+
+
+    if (wolfSSL_CryptHwMutexLock() == 0) { 
+        status = sss_digest_context_init(&digest_ctx, cfg_se050_i2c_pi, algo, kMode_SSS_Digest);  
+        if(status != kStatus_SSS_Success){
+            printf("error 1\n");
+            return -1;
```

Comment:
Please fix so HW mutex will unlock on error. Refactor for single return point at end so `wolfSSL_CryptHwMutexUnLock` is always called.